Add SLE BCI verification with Podman (#101)
SUSE/bci-docs: content/guides/verify-with-cosign.adoc

crypto/rsa: verification error
main.go:62: error during command execution: no matching signatures:
crypto/rsa: verification error
----

== Verifying SLE BCI with Podman
There are three main prerequisites to verify SLE Base Container Images using Podman. First, specify `registry.suse.com` as the registry for which image verification will be enabled.

=== Update Podman registries configuration
{{< hint type="note" >}}
Skip this step on SLE or openSUSE, as the correct configuration is already in place.
{{< /hint >}}

Add the following configuration to `/etc/containers/registries.d/default.yaml`:

[source,YAML]
----
docker:
  registry.suse.com:
    use-sigstore-attachments: true
----

Instead of editing the `default.yaml`, you can create a new file in
`/etc/containers/registries.d/` with a filename of your choice.

=== Update signature verification policy (`/etc/containers/policy.json`)

> Next, update the https://github.com/containers/image/blob/main/docs/containers-policy.json.5.md[policy.json] file.

Reviewer: I know the full path is in the title, but I still completely missed it, so better safe than sorry?
```suggestion
Next, update the https://github.com/containers/image/blob/main/docs/containers-policy.json.5.md[/etc/containers/policy.json] file.
```

Under the `docker` attribute, add the `registry.suse.com` configuration similar to the following:

[source,json]
----
{
    "default": [
        {
            "type": "insecureAcceptAnything"
        }
    ],
    "transports":
        {
        "docker-daemon":
            {
                "": [{"type":"insecureAcceptAnything"}]
            },
	    "docker":
		{
		    "registry.suse.com": [{
			   "type": "sigstoreSigned",
			   "keyPath": "/usr/share/pki/containers/suse-container-key.pem",
			   "signedIdentity": {"type": "matchRepository"}
			}]
		}
	}
}

> ----

Reviewer: It would be nice to mention here what to do, if there are already entries in the `transports.docker-daemon.docker` section/key. E.g. policy.json on Fedora looks like this:
```json
{
    "default": [
        {
            "type": "insecureAcceptAnything"
        }
    ],
    "transports": {
        "docker": {
            "registry.access.redhat.com": [
                {
                    "type": "signedBy",
                    "keyType": "GPGKeys",
                    "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release"
                }
            ],
            "registry.redhat.io": [
                {
                    "type": "signedBy",
                    "keyType": "GPGKeys",
                    "keyPath": "/etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release"
                }
            ]
        },
        "docker-daemon": {
            "": [
                {
                    "type": "insecureAcceptAnything"
                }
            ]
        }
    }
}
```
I would suggest to add an explanation what to do in this case.

Author: That's a good point, I've added a small note section clarifying this. PTAL.

crypto/rsa: verification error
main.go:62: error during command execution: no matching signatures:
crypto/rsa: verification error
----

== Verifying SLE BCI with Podman
There are three main prerequisites to verify SLE Base Container Images using Podman. First, specify `registry.suse.com` as the registry for which image verification will be enabled.

=== Update Podman registries configuration
{{< hint type="note" >}}

> This step can be skipped on SLE or openSUSE. The necessary configuration is already in place.

Reviewer 2: ```suggestion
Skip this step on SLE or openSUSE, as the correct configuration is already in place.
```

{{< /hint >}}

> Insert the following configuration into `/etc/containers/registries.d/default.yaml`:

Reviewer 2: ```suggestion
Add the following configuration to `/etc/containers/registries.d/default.yaml`:
```

[source,YAML]
----
docker:
  registry.suse.com:
    use-sigstore-attachments: true
----

> Instead of editing the `default.yaml`, you can choose to create a new file under
> `/etc/containers/registries.d/` with a filename of your choice.

Reviewer 2: ```suggestion
Instead of editing the `default.yaml`, you can create a new file in
`/etc/containers/registries.d/` with a filename of your choice.
```

=== Update signature verification policy (`/etc/containers/policy.json`)

> Next, you've to update the https://github.com/containers/image/blob/main/docs/containers-policy.json.5.md[policy.json] file.

Reviewer 2: ```suggestion
Next, update the https://github.com/containers/image/blob/main/docs/containers-policy.json.5.md[policy.json] file.
```

Under the `docker` attribute, add the `registry.suse.com` snippet similar to the following:

[source,json]
----
{
    "default": [
        {
            "type": "insecureAcceptAnything"
        }
    ],
    "transports":
        {
        "docker-daemon":
            {
                "": [{"type":"insecureAcceptAnything"}]
            },
	    "docker":
		{
		    "registry.suse.com": [{
			   "type": "sigstoreSigned",
			   "keyPath": "/usr/share/pki/containers/suse-container-key.pem",
			   "signedIdentity": {"type": "matchRepository"}
			}]
		}
	}
}
----

> You need to add the block `registry.suse.com` under `docker`. This instructs Podman,
> Skopeo and Buildah to verify images under the
> `registry.suse.com` repository. For instance, Podman, before pulling the image,
> would ensure that the signature is valid with respect to the public key we've
> specified above and reject the image otherwise.

Reviewer 2: ```suggestion
The specified configuration instructs Podman, Skopeo and Buildah to verify images under the
`registry.suse.com` repository. This way, before pulling the image, Podman checks the validity of the signature using the specified public key, and rejects the image if the validation fails.
```

=== Fetch the SUSE Container signing key
{{< hint type="note" >}}
This step is optional on SLE. The signing key is already installed under `/usr/share/pki/containers/suse-container-key.pem`.
{{< /hint >}}

> Fetch the public key used to sign SLE BCIs from https://www.suse.com/support/security/keys/[SUSE Signing Keys] or as shown below:

Reviewer 2: ```suggestion
Fetch the public key used to sign SLE BCIs from https://www.suse.com/support/security/keys/[SUSE Signing Keys], or use the following command:
```

[source,ShellSession]
----
> sudo curl -s https://ftp.suse.com/pub/projects/security/keys/container–key.pem \
    -o /usr/share/pki/containers/suse-container-key.pem
----

=== Verifying if the image is signed
Buildah, Podman and Skopeo will automatically verify every image pulled from `registry.suse.com` from now on. There are no additional steps required.

> Should image verification fail for some reason, then the respective command will fail with an error message like:

Reviewer 2: ```suggestion
If verification fails, the command returns an error message as follows:
```

[source,ShellSession]
----
> podman pull registry.suse.com/bci/bci-base:latest
Trying to pull registry.suse.com/bci/bci-base:latest...
Error: copying system image from manifest list: Source image rejected: Signature for identity registry.suse.com/bci/bci-base is not accepted
----

If there are no issues with the signed image and your configuration, you can continue with your

> usual development workflow.

Reviewer 2: This para is strictly speaking is not necessary. Perhaps a better ending would be to list the output that the command displays if verification was successful?

Author: So there's no change in output as such when you've set this kind of verification up. And that's why I thought keeping users aware of what to expect would be a nice good-to-have. What do you think?

Reviewer 2: Ah, OK. Makes sense to keep it, then.

crypto/rsa: verification error
main.go:62: error during command execution: no matching signatures:
crypto/rsa: verification error
----

== Verifying SLE BCI with Podman

> To verify a SLE BCI image with your Podman installation, there are three main
> pre-requisites as shown below. are required. First of all, you are required to
> specify  the registries you need to verify, `suse.registry.com` in this case.

Reviewer: ```suggestion
There are three main prerequisites to verify SLE Base Container Images using Podman. First, specify `registry.suse.com` as the registry for which image verification will be enabled.
```

=== Update podman registries configuration

> Update `/etc/containers/registries.d/default.yaml` with the following:

Reviewer: ```suggestion
{{< hint type="note" >}}
This step can be skipped on SLE or openSUSE. The necessary configuration is already in place.
{{< /hint >}}

Update `/etc/containers/registries.d/default.yaml` with the following:
```

> Update `/etc/containers/registries.d/default.yaml` with the following:

Reviewer: ```suggestion
Insert the following configuration into `/etc/containers/registries.d/default.yaml`:
```

> [source,ShellSession]

Reviewer: ```suggestion
[source,YAML]
```

----
docker:
  registry.suse.com:
    use-sigstore-attachments: true
----

> If you're using TW/SLE, you can skip this step since this configuration is already in place.

Reviewer: ```suggestion
```

Author: Isn't worth mentioning?

Instead of editing the `default.yaml`, you can choose to create a new file under
`/etc/containers/registries.d/` with a filename of your choice but make sure a single repository or

> namespace is not present in multiple files.

Reviewer: What do you mean by a repository or namespace here? We are talking about config files.

Author: I meant a repository (suse.registry.com) or a namespace within that repo (suse.registry.com/bci) shouldn't be duplicated elsewhere in `/etc/containers/registries.d/`

Reviewer: Honestly, I would suggest to remove this whole paragraph. This is a guide and imho guides should be to the point and leave everything not important out. Explaining what repositories & namespaces are would just inflate the guide and not add anything to the goal of this guide

Author: Fair enough, I maybe wanted to point out this potential issue but you're right, it's convoluting the guide for not-so-favorable gains. Thanks

Author: I think we should still retain the bit about being able to add new files under the directory and just get rid of the second sentence about duplication, wdyt?

Reviewer: The current version looks good to me, let's keep it :+1:

=== Update signature verification policy (`/etc/containers/policy.json`)

> Next, you've to update the https://github.com/containers/image/blob/main/docs/containers-policy.json.5.md[policy.json] file with the following contents:

Reviewer: ```suggestion
Update the https://github.com/containers/image/blob/main/docs/containers-policy.json.5.md[policy.json] file with the following contents:
```
Note, what does "update" mean here? Include? Replace?

Author: "Update" the policy file with the given contents.

Reviewer: Uh so what do I as the user have to do? Just copy and paste that snippet erasing everything else, or ensure that the other transport parts are still present?

Author: I think "modify" sounds better here, wdyt. The idea is to show what is going to happen and then in the second sentence, show how to do it.

Reviewer: Yes, that would indeed sound better. Good idea!

> [source,ShellSession]

Reviewer: ```suggestion
[source,json]
```

----
{
    "default": [
        {
            "type": "insecureAcceptAnything"
        }
    ],
    "transports":
        {
        "docker-daemon":
            {
                "": [{"type":"insecureAcceptAnything"}]
            },
	    "docker":
		{
		    "registry.suse.com": [{
			   "type": "sigstoreSigned",
			   "keyPath": "/usr/share/pki/containers/suse-container-key.pem",
			   "signedIdentity": {"type": "matchRepository"}
			}]
		}
	}
}
----
You need to add the block `registry.suse.com` under `docker`. This instructs podman,

> skopeo, buildah and other containers/image consumers to verify images under the

Reviewer: I would omit the "containers/image consumers". This is an end user documentation and end users have no idea what that means. Those who know what "containers/image consumers are, will know already.

Author: You're right, I don't know why I added that there.

`registry.suse.com` repository. For instance, podman, before pulling the image,
would ensure that the signature is valid with respect to the public key we've
specified above and reject the image otherwise.

=== Fetch the SUSE Container signing key

> This step is optional if you're on a SUSE distribution, in which case, this key should
> already be present under `/usr/share/pki/containers/suse-container-key.pem`. If not,
> you can fetch the public key used to sign SUSE BCIs from
> https://www.suse.com/support/security/keys/[SUSE Signing Keys]. Or you can follow
> the steps below:

Reviewer: ```suggestion
{{< hint type="note" >}}
This step is optional on SLE. The signing key is already installed under `/usr/share/pki/containers/suse-container-key.pem`.
{{< /hint >}}

Fetch the public key used to sign SLE BCIs from https://www.suse.com/support/security/keys/[SUSE Signing Keys] or as shown below:
```

[source,ShellSession]
----
> sudo curl -s https://ftp.suse.com/pub/projects/security/keys/container–key.pem \
    -o /usr/share/pki/containers/suse-container-key.pem
----

=== Verifying if the image is signed

> Once you have the pre-requisites in place, you don't have to explicitly verify each
> and every image from the configured repositories, `suse.registry.com` in this case.
> The respective tools, Podman here, should do that as and when you're dealing with
> container images from the said repository. For instance, if the verification fails for
> an image for whatever reason, you could expect the following output:

Reviewer: ```suggestion
Buildah, Podman and Skopeo will automatically verify every image pulled from `registry.suse.com` from now on. There are no additional steps required.

Should image verification fail for some reason, then the respective command will fail with an error message like:
```

> 

Reviewer: ```suggestion
[source,ShellSession]
```
